### PR TITLE
Copter: add Mission RTL mode and support DO_LAND_START

### PR DESCRIPTION
**Pull request by IamPete1**
This adds the ability to use DO_LAND_START's as a RTL option on copter, using the new Mission RTL mode. 

This also fixes the distance calculation of DO_LAND_START's with no location, we now use the next valid location.

More advanced features now broken out into a second PR that will follow this one.

~~This adds the option to do the shortest route to landing via DO_LAND_START's to both plane and copter.~~

~~This also adds a  new mission rejoin waypoint that allows the mission to be rejoined at any point on the return leg after the new waypoint.~~

~~Possibly this could be split up into two PRs,  the first adding the plane functionality to copter in the new mode, the second adding the new rejoin options to both plane and copter.~~

~~https://youtu.be/SoHXRTD0vEs~~

### CODE REVIEW

**Comment by IamPete1** [comment]
Just found https://github.com/ArduPilot/ardupilot/pull/11553 ......

File: libraries/AP_Mission/AP_Mission.h [hunk 1]
```diff
@@ -532,13 +532,21 @@ class AP_Mission
     // find the nearest landing sequence starting point (DO_LAND_START) and
     // return its index.  Returns 0 if no appropriate DO_LAND_START point can
     // be found.
-    uint16_t get_landing_sequence_start() const;
+    uint16_t get_landing_sequence_start();
+
+    // find the shortest distance to landing via DO_LAND_START and
+    // return its index.  Returns 0 if no appropriate DO_LAND_START point can
+    // be found.
+    uint16_t get_shortest_landing_sequence_start();
```
Comment (khancyr) [inline libraries/AP_Mission/AP_Mission.h:540]: Do you have an example use case for this ? I see it as a weak duplicate from the rally point, but I can be wrong

File: ArduCopter/Copter.h [hunk 2]
```diff
@@ -622,6 +622,13 @@ class Copter : public AP_Vehicle {
                                                       -1 // the priority list must end with a sentinel of -1
                                                      };
 
+    enum class RTL_AutoLand {
```
Comment (khancyr) [inline ArduCopter/Copter.h:625]: Maybe we could put this into AP_vehcile as that the same as plane.

File: libraries/AP_Mission/AP_Mission.h (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_Mission/AP_Mission.h:540]: You can only go directly to a rally point, this is much much more powerful, allowing take a pre-defined route to get there. The current code just jumps to the closest do land start, even if that might be preceding massive mission leg, this new code accounts for the mission after the do land start, not the just distance to get to point its-self.

Comment (khancyr) [reply libraries/AP_Mission/AP_Mission.h:540]: That is a much better introduction to what this do IMHO ;-P

**Review by rmackay9 — CHANGES_REQUESTED** [review]
As discussed on the dev call, it would be best if this is implemented as a new mode (MRTL?).  I'm worried that the pilot will be confused if they ask for RTL and instead get Auto.

I think @IamPete1 is going to try to implement this as a new flight mode.  I think you'll need to be careful of the interaction between auto and this new mode though.  Ideally I think they should have different state.  I don't want to over complicate things but what happens if the pilot switches from Auto to MRTL and then back to Auto?

**Comment by IamPete1** [comment]
https://youtu.be/SoHXRTD0vEs

**Comment by IamPete1** [comment]
I still need to add this to flight mode switch, RC options ect.

**Comment by IamPete1** [comment]
Add to Flight mode channel descriptions and added aux switch option

**Comment by IamPete1** [comment]
This now backs up the auto mission state on mode entry and puts it back again on mode exit. So auto mode is now unaffected by this new mode hijacking it.

**Comment by rmackay9** [comment]
The ability to backup and restore the mission state is great, thanks!

**Comment by IamPete1** [comment]
I have significantly reduced the scope of this to make it easier to get it in, this now only brings copter upto the same level as plane. Copter gets a new mode, auto state is backed up and restored. 

The new mission re-join way-point and options will come as a second PR. 

The only remaining slightly odd thing is that you can set the waypoint index and restart the mission while in Mission RTL mode, in that case it will behave just as auto, changing its current target. AUTO is unaffected as we will restore it back when exiting the mode. I think this is probably just a 'don't do that then' one, would be quite tricky to lock down that, and how far to take it, would you still be allowed to upload a new waypoint? And even if we get it to ignore were still doing the wrong thing because we should apply that reset to the backed up auto state so that when you return to true auto mode that change in mission index is honored.

**Comment by IamPete1** [comment]
fixed mode reason conflict

**Comment by IamPete1** [comment]
I should add this to the various fail safe options.

**Comment by IamPete1** [comment]
https://github.com/ArduPilot/ardupilot/pull/18075 is now better

**Pull request closed without merging**